core/server/api/v2/settings.js

```diff
@@ -126,18 +126,49 @@ module.exports = {
                 .then(() => {
                     const siteApp = require('../../web/site/app');
 
-                    try {
-                        return siteApp.reload();
-                    } catch (err) {
-                        // bring back backup, otherwise your Ghost blog is broken
+                    const bringBackValidRoutes = () => {
+                        urlService.resetGenerators({releaseResourcesOnly: true});
+
                         return fs.copy(backupRoutesPath, `${config.getContentPath('settings')}/routes.yaml`)
                             .then(() => {
                                 return siteApp.reload();
-                            })
-                            .then(() => {
+                            });
+                    };
+
+                    try {
+                        siteApp.reload();
+                    } catch (err) {
+                        return bringBackValidRoutes()
+                            .finally(() => {
                                 throw err;
                             });
                     }
+
+                    let tries = 0;
+
+                    function isBlogRunning() {
```

Comment:
Not a huge deal but with a recursive function like this it's better to keep the state of recursion local to the function, this eliminates any possibility of a bug because something else messed with the state. It also means you can use it more than once, so it can be extracted into a util.

e.g. `function isBlogRunning(tries = 0) {`

and `return isBlogRunning(tries + 1)`